From 108c5d6ba703b93f1b8fdd337b9cb1700d0709d6 Mon Sep 17 00:00:00 2001
From: Lifeng Zheng <zhenglifeng1@huawei.com>
Date: Fri, 13 Jun 2025 15:27:51 +0800
Subject: [PATCH 472/960] downstream: arm64: topology: Setup amu fie when cpu
 hotplugging

Amu fie was set up by a cpufreq policy notifier after the policy was
created. This caused some problems:

1. The cpus related to the same policy would all fail to set up amu fie if
one of them couldn't pass the freq_counters_valid() check.

2. The cpus fail to set up amu fie would never have a chance to set up
again.

When boot with maxcpu=1 restrict, the support amu flags of the offline cpus
would never be setup. After that, when cpufreq policy was being created,
the online cpu might set up amu fie fail because the other cpus related to
the same policy couldn't pass the freq_counters_valid() check. Hotplug the
offline cpus, since the policy was already created, amu_fie_setup() would
never be called again. All cpus couldn't setup amu fie in this situation.

After commit 1f023007f5e7 ("arm64/amu: Use capacity_ref_freq() to set AMU
ratio"), the max_freq stores in policy data is never needed when setting up
amu fie.  This indicates that the setting up of amu fie does not depend on
the policy any more. So each cpu can set up amu fie separately during
hotplug and the problems above will be solved.

Fixes: 1eb5dde674f5 ("cpufreq: CPPC: Add support for frequency invariance")
Signed-off-by: Lifeng Zheng <zhenglifeng1@huawei.com>
Signed-off-by: Hongye Lin <linhongye@h-partners.com>
Signed-off-by: chenyi <chenyi211@h-partners.com>
Conflicts:
	arch/arm64/kernel/topology.c
[ Context conflicts. ]
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 arch/arm64/kernel/topology.c | 52 +++++++++++-------------------------
 1 file changed, 16 insertions(+), 36 deletions(-)

diff --git a/arch/arm64/kernel/topology.c b/arch/arm64/kernel/topology.c
index e5498242bc1d..071e538408bf 100644
--- a/arch/arm64/kernel/topology.c
+++ b/arch/arm64/kernel/topology.c
@@ -351,52 +351,28 @@ int arch_freq_get_on_cpu(int cpu)
 	return freq;
 }
 
-static void amu_fie_setup(const struct cpumask *cpus)
+static void amu_fie_setup(unsigned int cpu)
 {
-	int cpu;
-
-	/* We are already set since the last insmod of cpufreq driver */
-	if (unlikely(cpumask_subset(cpus, amu_fie_cpus)))
+	if (cpumask_test_cpu(cpu, amu_fie_cpus))
 		return;
 
-	for_each_cpu(cpu, cpus) {
-		if (!freq_counters_valid(cpu))
-			return;
-	}
+	if (!freq_counters_valid(cpu))
+		return;
 
-	cpumask_or(amu_fie_cpus, amu_fie_cpus, cpus);
+	cpumask_set_cpu(cpu, amu_fie_cpus);
 
 	topology_set_scale_freq_source(&amu_sfd, amu_fie_cpus);
 
-	pr_debug("CPUs[%*pbl]: counters will be used for FIE.",
-		 cpumask_pr_args(cpus));
+	pr_debug("CPUs[%u]: counters will be used for FIE.", cpu);
 }
 
-static int init_amu_fie_callback(struct notifier_block *nb, unsigned long val,
-				 void *data)
+static int cpuhp_topology_online(unsigned int cpu)
 {
-	struct cpufreq_policy *policy = data;
-
-	if (val == CPUFREQ_CREATE_POLICY)
-		amu_fie_setup(policy->related_cpus);
-
-	/*
-	 * We don't need to handle CPUFREQ_REMOVE_POLICY event as the AMU
-	 * counters don't have any dependency on cpufreq driver once we have
-	 * initialized AMU support and enabled invariance. The AMU counters will
-	 * keep on working just fine in the absence of the cpufreq driver, and
-	 * for the CPUs for which there are no counters available, the last set
-	 * value of arch_freq_scale will remain valid as that is the frequency
-	 * those CPUs are running at.
-	 */
+	amu_fie_setup(cpu);
 
 	return 0;
 }
 
-static struct notifier_block init_amu_fie_notifier = {
-	.notifier_call = init_amu_fie_callback,
-};
-
 static int __init init_amu_fie(void)
 {
 	int ret;
@@ -404,12 +380,16 @@ static int __init init_amu_fie(void)
 	if (!zalloc_cpumask_var(&amu_fie_cpus, GFP_KERNEL))
 		return -ENOMEM;
 
-	ret = cpufreq_register_notifier(&init_amu_fie_notifier,
-					CPUFREQ_POLICY_NOTIFIER);
-	if (ret)
+	ret = cpuhp_setup_state(CPUHP_AP_ONLINE_DYN,
+				"arm64/topology:online",
+				cpuhp_topology_online,
+				NULL);
+	if (ret < 0) {
 		free_cpumask_var(amu_fie_cpus);
+		return ret;
+	}
 
-	return ret;
+	return 0;
 }
 core_initcall(init_amu_fie);
 
-- 
2.43.0

